packages/pyright-internal/src/languageServerBase.ts

```diff
@@ -1061,6 +1096,85 @@ export abstract class LanguageServerBase implements LanguageServerInterface, Dis
         this.openFileMap.delete(uri.key);
     }
 
+    protected async onDiagnostics(params: DocumentDiagnosticParams, token: CancellationToken) {
+        const uri = this.convertLspUriStringToUri(params.textDocument.uri);
+        const workspace = await this.getWorkspaceForFile(uri);
+        let sourceFile = workspace.service.getSourceFile(uri);
+        let diagnosticsVersion = sourceFile?.isCheckingRequired()
+            ? UncomputedDiagnosticsVersion
+            : sourceFile?.getDiagnosticVersion() ?? UncomputedDiagnosticsVersion;
+        const result: DocumentDiagnosticReport = {
+            kind: 'full',
+            resultId: sourceFile?.getDiagnosticVersion()?.toString(),
+            items: [],
+        };
+        if (
+            workspace.disableLanguageServices ||
+            !canNavigateToFile(workspace.service.fs, uri) ||
+            token.isCancellationRequested
+        ) {
+            return result;
+        }
+
+        // Reanalyze the file if it's not up to date.
+        if (params.previousResultId !== diagnosticsVersion.toString() && sourceFile) {
+            let diagnosticsVersionAfter = UncomputedDiagnosticsVersion - 1; // Just has to be different
+            let serverDiagnostics: AnalyzerDiagnostic[] = [];
+
+            // Loop until we analyze the same version that we started with.
+            while (diagnosticsVersion !== diagnosticsVersionAfter && !token.isCancellationRequested) {
+                // Reset the version we're analyzing
+                sourceFile = workspace.service.getSourceFile(uri);
+                diagnosticsVersion = sourceFile?.getDiagnosticVersion() ?? UncomputedDiagnosticsVersion;
+
+                // Then reanalyze the file (this should go to the background thread so this thread can handle other requests).
+                await workspace.service.analyzeFile(uri, token);
+
+                // Then pick up the diagnostics created.
+                serverDiagnostics = sourceFile
+                    ? await workspace.service.getDiagnosticsForRange(uri, sourceFile.getRange(), token)
+                    : [];
+
+                // If any text edits came in, make sure we reanalyze the file.
+                const sourceFileAfter = workspace.service.getSourceFile(uri);
+                diagnosticsVersionAfter = sourceFileAfter?.getDiagnosticVersion() ?? UncomputedDiagnosticsVersion;
+            }
+
+            // Then convert the diagnostics to the LSP format.
+            const lspDiagnostics = this._convertDiagnostics(workspace.service.fs, serverDiagnostics).filter(
+                (d) => d !== undefined
+            ) as Diagnostic[];
+
+            result.resultId = diagnosticsVersion.toString();
```

Comment:
Should this be `diagnosticsVersionAfter`?

Reply:
I don't think so. They're the same as long as the loop exits.

Reply:
Ah, right, of course. For what it's worth, I found this loop extremely confusing. I stared at it for a long time trying to figure out what was going on here. I don't have specific recommendations on how to improve it unfortunately. There was some similar code that I reviewed in pyrx recently that I also found very difficult to wrap my head around. Might just be me.